client/src/test/java/io/pravega/client/segment/impl/SegmentOutputStreamTest.java

```diff
@@ -213,21 +207,98 @@ public void testFlush() throws ConnectionFailedException, SegmentSealedException
     }
 
     @Test
-    @Ignore
-    public void testAutoClose() {
-        fail();
-    }
+    public void testConnectionFailure() throws ConnectionFailedException {
+        UUID cid = UUID.randomUUID();
+        PravegaNodeUri uri = new PravegaNodeUri("endpoint", SERVICE_PORT);
+        MockConnectionFactoryImpl cf = new MockConnectionFactoryImpl(uri);
+        MockController controller = new MockController(uri.getEndpoint(), uri.getPort(), cf);
+        ClientConnection connection = mock(ClientConnection.class);
+        cf.provideConnection(uri, connection);
+        SegmentOutputStreamImpl output = new SegmentOutputStreamImpl(SEGMENT, controller, cf, cid);
+        output.setupConnection();
+        InOrder inOrder = Mockito.inOrder(connection);
+        inOrder.verify(connection).send(new WireCommands.SetupAppend(1, cid, SEGMENT));
+        cf.getProcessor(uri).appendSetup(new WireCommands.AppendSetup(1, SEGMENT, cid, 0));
+        ByteBuffer data = getBuffer("test");
 
-    @Test
-    @Ignore
-    public void testFailOnAutoClose() {
-        fail();
+        CompletableFuture<Boolean> acked = new CompletableFuture<>();
+        Append append = new Append(SEGMENT, cid, 1, Unpooled.wrappedBuffer(data), null);
+        Mockito.doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                cf.getProcessor(uri).connectionDropped();
+                throw new ConnectionFailedException();
+            }
+        }).when(connection).send(append);
+        Mockito.doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                CompletedCallback callback = (CompletedCallback) invocation.getArgument(1);
+                callback.complete(null);
+                return null;
+            }
+        }).when(connection).sendAsync(Mockito.eq(Collections.singletonList(append)), Mockito.any());
+        
+        Async.testBlocking(() -> {            
+            output.write(new PendingEvent(null, data, acked, null));
+        }, () -> {            
+            cf.getProcessor(uri).appendSetup(new WireCommands.AppendSetup(2, SEGMENT, cid, 0));
+        });
+        inOrder.verify(connection).send(append);
+        inOrder.verify(connection).send(new WireCommands.SetupAppend(2, cid, SEGMENT));
+        inOrder.verify(connection).sendAsync(Mockito.eq(Collections.singletonList(append)), Mockito.any());
+        assertEquals(false, acked.isDone());
+        inOrder.verifyNoMoreInteractions();
     }
 
+    
     @Test
-    @Ignore
-    public void testOutOfOrderAcks() {
-        fail();
+    public void testFailOnClose() throws ConnectionFailedException, SegmentSealedException {
```

Comment:
This test case seems to be doing the right thing, but I find it hard to read. It is not strictly necessary, but adding comments that help guide the code reader, makes it easier to determine whether the behavior is correct. I suppose the other argument could be made to, that not adding any comment forces the code reviewer to understand the code and verify that it works. There is probably some middle ground....